app/src/main/java/com/keylesspalace/tusky/components/compose/view/ProgressImageView.kt

```diff
@@ -0,0 +1,115 @@
+import android.graphics.PorterDuffXfermode
+import android.graphics.RectF
+import android.graphics.drawable.Drawable
+import android.util.AttributeSet
+import androidx.appcompat.content.res.AppCompatResources
+import at.connyduck.sparkbutton.helpers.Utils
+import com.keylesspalace.tusky.R
+import com.keylesspalace.tusky.view.MediaPreviewImageView
+
+class ProgressImageView : MediaPreviewImageView {
+    private var progress = -1
+    private val progressRect = RectF()
+    private val biggerRect = RectF()
+    private val circlePaint = Paint(Paint.ANTI_ALIAS_FLAG)
+    private val clearPaint = Paint(Paint.ANTI_ALIAS_FLAG)
+    private val markBgPaint = Paint(Paint.ANTI_ALIAS_FLAG)
+    private lateinit var captionDrawable: Drawable
+
+    constructor(context: Context?) : super(context!!) {
+        init()
+    }
+
+    constructor(context: Context?, attrs: AttributeSet?) : super(
+        context!!,
+        attrs
+    ) {
+        init()
+    }
+
+    constructor(context: Context?, attrs: AttributeSet?, defStyleAttr: Int) : super(
+        context!!,
+        attrs,
+        defStyleAttr
+    ) {
+        init()
+    }
+
+    private fun init() {
+        circlePaint.color = context.getColor(R.color.tusky_blue)
+        circlePaint.strokeWidth = Utils.dpToPx(context, 4).toFloat()
+        circlePaint.style = Paint.Style.STROKE
+        clearPaint.xfermode = PorterDuffXfermode(PorterDuff.Mode.DST_OUT)
+        markBgPaint.style = Paint.Style.FILL
+        markBgPaint.color = context.getColor(R.color.tusky_grey_10)
+        captionDrawable = AppCompatResources.getDrawable(context, R.drawable.spellcheck)!!
+    }
+
+    fun setProgress(progress: Int) {
+        this.progress = progress
+        if (progress != -1) {
+            setColorFilter(Color.rgb(123, 123, 123), PorterDuff.Mode.MULTIPLY)
+        } else {
+            clearColorFilter()
+        }
+        invalidate()
+    }
+
+    fun setChecked(checked: Boolean) {
+        markBgPaint.color =
+            context.getColor(if (checked) R.color.tusky_blue else R.color.tusky_grey_10)
+        invalidate()
+    }
+
+    override fun onDraw(canvas: Canvas) {
+        super.onDraw(canvas)
+        val angle = progress / 100f * 360 - 90
+        val halfWidth = width / 2f
+        val halfHeight = height / 2f
+        progressRect[halfWidth * 0.75f, halfHeight * 0.75f, halfWidth * 1.25f] = halfHeight * 1.25f
+        biggerRect.set(progressRect)
+        val margin = 8
+        biggerRect[progressRect.left - margin, progressRect.top - margin, progressRect.right + margin] =
+            progressRect.bottom + margin
+        canvas.saveLayer(biggerRect, null, Canvas.ALL_SAVE_FLAG)
+        if (progress != -1) {
+            canvas.drawOval(progressRect, circlePaint)
+            canvas.drawArc(biggerRect, angle, 360 - angle - 90, true, clearPaint)
+        }
+        canvas.restore()
+        val circleRadius = Utils.dpToPx(context, 14)
+        val circleMargin = Utils.dpToPx(context, 14)
```

Comment:
Calling this twice inside onDraw despite it always retuning the same does not look like a good idea. How about some class member fourteenDpInPx and then a comment explaining it is used both as radius and as margin?
(`onDraw` should really be as minimal as possible for perf reasons, it is called very often)